data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/S3SelectObjectWorker.java

```diff
@@ -191,6 +234,44 @@ private void parseCompleteStreamFromResponseHeader(final S3ObjectReference s3Obj
         }
     }
 
+    private boolean flushWithBackoff(final BufferAccumulator<Record<Event>> bufferAccumulator, final int maxRetries, final Duration initialDelay) {
+
+        final ScheduledExecutorService scheduledExecutorService = Executors.newSingleThreadScheduledExecutor();
+        long nextDelay = initialDelay.toMillis();
+        int backoffRate = 2;
+        boolean flushedSuccessfully;
+
+        for (int retryCount = 0; retryCount < maxRetries; retryCount++) {
+            final ScheduledFuture<Boolean> flushBufferFuture = scheduledExecutorService.schedule(() -> {
+                try {
+                    bufferAccumulator.flush();
+                    return true;
+                } catch (final Exception e) {
+                    return false;
+                }
+            }, nextDelay, TimeUnit.MILLISECONDS);
+
+            try {
+                flushedSuccessfully = flushBufferFuture.get();
+                if (!flushedSuccessfully) {
+                    nextDelay *= backoffRate;
+                } else {
+                    scheduledExecutorService.shutdownNow();
+                    return true;
+                }
+            } catch (ExecutionException e) {
+                LOG.warn("Retrying of flushing the buffer accumulator hit an exception: {}", e.getMessage());
+            } catch (InterruptedException e) {
+                LOG.warn("Retrying of flushing the buffer accumulator was interrupted: {}", e.getMessage());
+            }
+        }
+
+
+        LOG.warn("Flushing the bufferAccumulator failed after {} attempts", maxRetries);
```

Comment:
We should back off and retry indefinitely here IMO. If the current docs in the accumulator cannot be written to the buffer due to a timeout, pulling in more data doesn't seem like the correct behavior.

If a sink is unhealthy and the buffer is never clearing, we essentially stop the pipeline via infinite backoff and retry. Once the sink is healed, the process would pick back up where it left off without losing data or going OOM from the Accumulator continuing to accept data.

If we do go with infinite retry, we may want to use a linear backoff for responsiveness.

Reply:
Makes sense. Can I keep this easily changeable by setting the retryCount to INTEGER.MAX_VALUE?

Reply:
That works for me. I think the OpenSearch sink does the same thing with `Integer.MAX_VALUE`